source/base/tensor_function_parser.cc

```diff
@@ -0,0 +1,419 @@
+  // we expect to read in.  We
+  // distinguish between two cases:
+  // Time dependent problems, and not
+  // time dependent problems. In the
+  // first case the number of
+  // variables is given by the
+  // dimension plus one. In the other
+  // case, the number of variables is
+  // equal to the dimension. Once we
+  // parsed the variables string, if
+  // none of this is the case, then
+  // an exception is thrown.
+  if (time_dependent)
+    n_vars = dim + 1;
+  else
+    n_vars = dim;
+
+  // create a parser object for the current thread we can then query
+  // in value() and vector_value(). this is not strictly necessary
+  // because a user may never call these functions on the current
+  // thread, but it gets us error messages about wrong formulas right
+  // away
+  init_muparser();
+
+  // finally set the initialization bit
+  initialized = true;
+}
+
+
+
+template <int rank, int dim, typename Number>
+void
+TensorFunctionParser<rank, dim, Number>::init_muparser() const
+{
+  // check that we have not already initialized the parser on the
+  // current thread, i.e., that the current function is only called
+  // once per thread
+  Assert(tfp.get().size() == 0, ExcInternalError());
+
+  // initialize the objects for the current thread (tfp.get() and
+  // vars.get())
+  tfp.get().reserve(this->n_components);
+  vars.get().resize(var_names.size());
+  for (unsigned int component = 0; component < this->n_components; ++component)
+    {
+      tfp.get().emplace_back(new mu::Parser());
+
+      for (std::map<std::string, double>::const_iterator constant =
+             constants.begin();
+           constant != constants.end();
+           ++constant)
+        {
+          tfp.get()[component]->DefineConst(constant->first, constant->second);
+        }
+
+      for (unsigned int iv = 0; iv < var_names.size(); ++iv)
+        tfp.get()[component]->DefineVar(var_names[iv], &vars.get()[iv]);
+
+      // define some compatibility functions:
+      tfp.get()[component]->DefineFun("if", internal::mu_if, true);
+      tfp.get()[component]->DefineOprt("|", internal::mu_or, 1);
+      tfp.get()[component]->DefineOprt("&", internal::mu_and, 2);
+      tfp.get()[component]->DefineFun("int", internal::mu_int, true);
+      tfp.get()[component]->DefineFun("ceil", internal::mu_ceil, true);
+      tfp.get()[component]->DefineFun("cot", internal::mu_cot, true);
+      tfp.get()[component]->DefineFun("csc", internal::mu_csc, true);
+      tfp.get()[component]->DefineFun("floor", internal::mu_floor, true);
+      tfp.get()[component]->DefineFun("sec", internal::mu_sec, true);
+      tfp.get()[component]->DefineFun("log", internal::mu_log, true);
+      tfp.get()[component]->DefineFun("pow", internal::mu_pow, true);
+      tfp.get()[component]->DefineFun("erfc", internal::mu_erfc, true);
+      tfp.get()[component]->DefineFun("rand_seed", internal::mu_rand_seed, true);
+      tfp.get()[component]->DefineFun("rand", internal::mu_rand, true);
```

Comment:
Could we put a function that does this inside the internal muparser header?

Reply:
Done

Reply:
Can that be taken care of together with the clean up of function_parser.cc/h?
See issue #9143